### PR TITLE
YJDH-62: Shared Wizard component allowing dynamic steps + example in kesaseteli

### PR DESCRIPTION
**Pull request by nenonja**
## Description :sparkles:
Common Wizard component into shared folder. It is based on https://github.com/devrnt/react-use-wizard but I didnt want to use the original package because it had some limitations. Wizard allows dynamic steps added as child components like this:
```
<Wizard header={WizardHeader} footer={WizardFooter}>
     <Step1 />
     <Step2 />
     <ThirdStep>
         Hello Step 3
     </ThirdStep>   
     ....
 </Wizard>
 ```
 It does _not_ use nextjs routes so it's more reusable. also `useWizard`-hook can be used in steps and wizard header and footer to get the wizard functionality available (go to previous/next step, read current state number, read number of steps etc)

There is several components copy-pasted from benefit to shared folder:
* ErrorPage
* PageLoadingSpinner
* Toast
* WizardStepper (same as old Stepper but uses new Wizard's state + clickable steps)

I think we can assume that every project that uses shared components uses also next-i18next translations. Then there is no need to prop drill every translation text to the shared component but just add the translation to the json file. If it's ok then benefit could start to use same shared components and delete the original copy-pasted component from benefit's folder.

### CODE REVIEW

**Comment by nenonja** [comment]
> SonarCloud Quality Gate failed.    Quality Gate failed
> 
> [Bug](https://sonarcloud.io/project/issues?id=City-of-Helsinki_yjdh&pullRequest=204&resolved=false&types=BUG) [A](https://sonarcloud.io/project/issues?id=City-of-Helsinki_yjdh&pullRequest=204&resolved=false&types=BUG) [0 Bugs](https://sonarcloud.io/project/issues?id=City-of-Helsinki_yjdh&pullRequest=204&resolved=false&types=BUG)
> [Vulnerability](https://sonarcloud.io/project/issues?id=City-of-Helsinki_yjdh&pullRequest=204&resolved=false&types=VULNERABILITY) [A](https://sonarcloud.io/project/issues?id=City-of-Helsinki_yjdh&pullRequest=204&resolved=false&types=VULNERABILITY) [0 Vulnerabilities](https://sonarcloud.io/project/issues?id=City-of-Helsinki_yjdh&pullRequest=204&resolved=false&types=VULNERABILITY)
> [Security Hotspot](https://sonarcloud.io/project/security_hotspots?id=City-of-Helsinki_yjdh&pullRequest=204&resolved=false&types=SECURITY_HOTSPOT) [A](https://sonarcloud.io/project/security_hotspots?id=City-of-Helsinki_yjdh&pullRequest=204&resolved=false&types=SECURITY_HOTSPOT) [0 Security Hotspots](https://sonarcloud.io/project/security_hotspots?id=City-of-Helsinki_yjdh&pullRequest=204&resolved=false&types=SECURITY_HOTSPOT)
> [Code Smell](https://sonarcloud.io/project/issues?id=City-of-Helsinki_yjdh&pullRequest=204&resolved=false&types=CODE_SMELL) [A](https://sonarcloud.io/project/issues?id=City-of-Helsinki_yjdh&pullRequest=204&resolved=false&types=CODE_SMELL) [8 Code Smells](https://sonarcloud.io/project/issues?id=City-of-Helsinki_yjdh&pullRequest=204&resolved=false&types=CODE_SMELL)
> 
> [No Coverage information](https://sonarcloud.io/component_measures?id=City-of-Helsinki_yjdh&pullRequest=204) No Coverage information
> [](https://sonarcloud.io/component_measures?id=City-of-Helsinki_yjdh&pullRequest=204&metric=new_duplicated_lines_density&view=list) [5.1% Duplication](https://sonarcloud.io/component_measures?id=City-of-Helsinki_yjdh&pullRequest=204&metric=new_duplicated_lines_density&view=list)

Lots of duplication due to these copy-pasted components from benefit to shared which can be later deleted from benefit code.

File: frontend/kesaseteli/employer/src/__tests__/utils/components/render-component.tsx [hunk 1]
```diff
@@ -13,7 +16,10 @@ const renderComponent = (
 ): RenderResult =>
   render(
     <BackendAPIContext.Provider value={AxiosTestContext}>
-      <QueryClientProvider client={client}>{Component}</QueryClientProvider>
+      <QueryClientProvider client={client}>
+        <ThemeProvider theme={theme}>{Component}</ThemeProvider>
+        <HiddenLoadingIndicator />
```
Comment (nenonja) [inline frontend/kesaseteli/employer/src/__tests__/utils/components/render-component.tsx:21]: Helper indicator for the tests which will tell if some backend request is till pending

File: frontend/kesaseteli/employer/src/__tests__/utils/components/render-page.tsx [hunk 2]
```diff
@@ -32,6 +34,7 @@ const renderPage = (
             </Layout>
           </ThemeProvider>
         </AuthProvider>
+        <HiddenLoadingIndicator />
```
Comment (nenonja) [inline frontend/kesaseteli/employer/src/__tests__/utils/components/render-page.tsx:37]: Helper indicator for the tests which will tell if some backend request is till pending

File: frontend/kesaseteli/employer/src/pages/_app.tsx [hunk 3]
```diff
@@ -44,6 +45,7 @@ const App: React.FC<AppProps> = ({ Component, pageProps }) => (
           </Layout>
         </ThemeProvider>
       </AuthProvider>
+      <HiddenLoadingIndicator />
```
Comment (nenonja) [inline frontend/kesaseteli/employer/src/pages/_app.tsx:48]: Also added this hidden loading indicator to the _app for helping the browser tests to check that all backend requests are done

File: frontend/kesaseteli/employer/src/pages/login.tsx [hunk 4]
```diff
@@ -1,23 +1,27 @@
...
-        <p>Käyttäjäsessio vanhentui. Kirjaudu uudelleen sisään.</p>
-      )}
-    </Layout>
+    <Container>
+      <Layout headingText="Työnantajan liittymä">
+        {logout && <p>Olet kirjautunut ulos</p>}
+        {error && <p>Tapahtui tuntematon virhe. Kirjaudu uudelleen sisään.</p>}
+        {sessionExpired && (
+          <p>Käyttäjäsessio vanhentui. Kirjaudu uudelleen sisään.</p>
+        )}
```
Comment (Nizar-Rahme) [inline frontend/kesaseteli/employer/src/pages/login.tsx:22]: Are these texts left out of the translation files on purpose?

File: frontend/kesaseteli/employer/public/locales/en/common.json [hunk 5]
```diff
@@ -1,13 +1,25 @@
...
+  },
   "application": {
+    "new": "New application",
+    "common_error": "Unknown error occurred: ",
+    "loading": "Loading application...",
+    "loadingFinished": "Loading application has finished",
     "step1": {
+      "name": "Employer",
       "header": "1. Employer information",
+      "tooltip": "EN Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.",
```
[the remark this answers is not in the document]
Comment (nenonja) [reply frontend/kesaseteli/employer/public/locales/en/common.json:22]: Yes tooltip could be easily forgotten but we going to have a process that final english and swedish translations will come from excel file so all these values will be checked, now these are just unofficial translations made by developers :)

File: frontend/kesaseteli/employer/src/components/application/form/ActionButtons.sc.ts [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+  border-color: ${(props) => props.theme.colors.black90} !important;
+  border-width: 3px !important;
+  min-width: 170px;
+  max-height: 60px;
+`;
+
+export const $PrimaryButton = styled(Button)`
+  border-width: 3px !important;
+  width: 255px;
+`;
```
[the remark this answers is not in the document]
Comment (nenonja) [reply frontend/kesaseteli/employer/src/components/application/form/ActionButtons.sc.ts:15]: That's great thing! These are copy-pasted from benefit's

File: frontend/kesaseteli/employer/src/pages/login.tsx (hunk 4 above)
Comment (nenonja) [reply frontend/kesaseteli/employer/src/pages/login.tsx:22]: Yes this login page is very old and need to be finalized before release

File: frontend/kesaseteli/employer/src/components/application/steps/step1/Step1Employer.tsx [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+            validation={{
+              required: true,
+              maxLength: 254, // eslint-disable-next-line security/detect-unsafe-regex
+              pattern: /^(([^\s"(),.:;<>@[\\\]]+(\.[^\s"(),.:;<>@[\\\]]+)*)|(".+"))@((\[(?:\d{1,3}\.){3}\d{1,3}])|(([\dA-Za-z-]+\.)+[A-Za-z]{2,}))$/,
+            }}
+          />
+          <TextInput
+            id="invoicer_phone_number"
+            validation={{ required: true, maxLength: 64 }}
+          />
```
Comment (Nizar-Rahme) [inline frontend/kesaseteli/employer/src/components/application/steps/step1/Step1Employer.tsx:46]: Having the validation inline here is error-prone, and makes testing hard.

File: frontend/kesaseteli/employer/src/components/application/steps/step1/Step1Employer.tsx [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+          <TextInput
+            id="invoicer_name"
+            validation={{ required: true, maxLength: 256 }}
+          />
+          <TextInput
+            id="invoicer_email"
+            validation={{
+              required: true,
+              maxLength: 254, // eslint-disable-next-line security/detect-unsafe-regex
+              pattern: /^(([^\s"(),.:;<>@[\\\]]+(\.[^\s"(),.:;<>@[\\\]]+)*)|(".+"))@((\[(?:\d{1,3}\.){3}\d{1,3}])|(([\dA-Za-z-]+\.)+[A-Za-z]{2,}))$/,
```
Comment (Nizar-Rahme) [inline frontend/kesaseteli/employer/src/components/application/steps/step1/Step1Employer.tsx:40]: `Yup` has a built-in method for validating emails.

File: frontend/kesaseteli/employer/src/hooks/backend/useOperationPermitted.ts [hunk 9]
```diff
@@ -0,0 +1,12 @@
+import useLogoutQuery from 'kesaseteli/employer/hooks/backend/useLogoutQuery';
+import useAuth from 'shared/hooks/useAuth';
+
+const useIsOperationPermitted = (): boolean => {
+  const { isAuthenticated } = useAuth();
+  const {
+    isLoading: isLoadingLogout,
+    isSuccess: isLoggedOut,
+  } = useLogoutQuery();
+  return isAuthenticated && !isLoadingLogout && !isLoggedOut;
```
Comment (Nizar-Rahme) [inline frontend/kesaseteli/employer/src/hooks/backend/useOperationPermitted.ts:10]: Is there a way that the user would be logged out AND authenticated?! I think `!isLoggedOut` seems redundant.

File: frontend/shared/src/components/stepper/Stepper.tsx [hunk 10]
```diff
@@ -18,10 +18,8 @@ const Stepper: React.FC<StepperProps> = ({ activeStep = 1, steps }) => (
   </$StepsContainer>
 );
 
-const defaultProps = {
+Stepper.defaultProps = {
   activeStep: 1,
 };
 
-Stepper.defaultProps = defaultProps;
-
```
Comment (Nizar-Rahme) [inline frontend/shared/src/components/stepper/Stepper.tsx:26]: I think we should get rid of these `defaultProps`, this `({ activeStep = 1, steps }) => (` would do the trick.

File: frontend/shared/src/components/hocs/withAuthRedirect.tsx [hunk 11]
```diff
@@ -1,10 +1,11 @@
 import { useRouter } from 'next/router';
 import React from 'react';
+import PageLoadingSpinner from 'shared/components/pages/PageLoadingSpinner';
 import useAuth from 'shared/hooks/useAuth';
 import isServerSide from 'shared/server/is-server-side';
 
 const DefaultLoadingFallback = (): React.ReactElement<unknown> => (
-  <p>Uudelleenohjataan...</p>
+  <PageLoadingSpinner />
 );
```
Comment (Nizar-Rahme) [inline frontend/shared/src/components/hocs/withAuthRedirect.tsx:9]: What's the purpose of this component? 
It looks like `DefaultLoadingFallback` is assigned to `LoadingComponent = DefaultLoadingFallback`; why don't you assign `LoadingComponent = PageLoadingSpinner` instead?

File: frontend/kesaseteli/employer/src/components/application/steps/step1/Step1Employer.tsx (hunk 8 above)
Comment (nenonja) [reply frontend/kesaseteli/employer/src/components/application/steps/step1/Step1Employer.tsx:40]: Including yup seem to add a lot of boilerplate: https://react-hook-form.com/advanced-usage#CustomHookwithResolver compared to react-hook-form builtin validation.
But these validation objects could be moved to own file. But I rather do it when meaking step2 if I have to start copy-pasting stuff.

File: frontend/kesaseteli/employer/src/components/application/steps/step1/Step1Employer.tsx (hunk 7 above)
Comment (nenonja) [reply frontend/kesaseteli/employer/src/components/application/steps/step1/Step1Employer.tsx:46]: There is already validation tests in application.test.tsx [here](https://github.com/City-of-Helsinki/yjdh/pull/204/files#diff-6c92d4860f92954993b55ba77efb5f9062292779297df9293651056f441cd6beR102) and [here](https://github.com/City-of-Helsinki/yjdh/pull/204/files#diff-6c92d4860f92954993b55ba77efb5f9062292779297df9293651056f441cd6beR78)

if you mean unit tests, they should be written for email regex and other upcoming regexes. In kesaseteli project the validations will be much more relaxed compared to benefit's so mostly min and max lengths only which are determined by database.
I'm planning to add email and other regex unit tests soon when there will be more validation related stuff coming with step 2.

File: frontend/kesaseteli/employer/src/hooks/backend/useOperationPermitted.ts (hunk 9 above)
Comment (nenonja) [reply frontend/kesaseteli/employer/src/hooks/backend/useOperationPermitted.ts:10]: I'm not sure if it's completely same. I can rename `isLoggedOut` to `isLogoutSucceeded`. There might be a tiny moment between the logout response returns successful result and `isAuthenticated` is false. The latter is evaluated after some authorized content is rendered and `withAuth` hoc has called backend to retrieve user data and response is 403.

The reason I added this hook when I realized that there will be some extra page renders with extra backend calls while clicking the logout button (that also caused lots of problems with component tests). Other way to tackle the problem is to somehow prevent extra renders when clicking logout button.

Comment (nenonja) [reply frontend/kesaseteli/employer/src/hooks/backend/useOperationPermitted.ts:10]: Also I dont know what happens if logout responses error, is user still authenticated or not 🤷

File: frontend/shared/src/components/stepper/Stepper.tsx (hunk 10 above)
Comment (nenonja) [reply frontend/shared/src/components/stepper/Stepper.tsx:26]: Let's try if I can remove it. I think that eslint complained about when not having it.

File: frontend/shared/src/components/hocs/withAuthRedirect.tsx (hunk 11 above)
Comment (nenonja) [reply frontend/shared/src/components/hocs/withAuthRedirect.tsx:9]: Definitely, coding too fast without thinking :)

File: frontend/shared/src/components/stepper/Stepper.tsx (hunk 10 above)
Comment (nenonja) [reply frontend/shared/src/components/stepper/Stepper.tsx]: seem to work!

File: frontend/shared/src/components/hocs/withAuthRedirect.tsx (hunk 11 above)
Comment (nenonja) [reply frontend/shared/src/components/hocs/withAuthRedirect.tsx]: fixed

**Review by melniiv — APPROVED** [review]

**Pull request merged**